#include "dinnerselection.h"
#include "ui_dinnerselection.h"
#include <QUrlQuery>
#include <QJsonDocument>
#include <QJsonObject>
#include <QJsonArray>
#include <QDebug>
#include <QRandomGenerator>
#include <QTimer>
#include <QMessageBox>
#include <QInputDialog>
#include <QFormLayout>
#include <QDialogButtonBox>
#include <QLineEdit>
#include <cmath>

DinnerSelection::DinnerSelection(QWidget *parent)
    : QWidget(parent)
    , ui(new Ui::DinnerSelection)
{
    setFixedSize(1200, 800);
    setMinimumSize(600, 500);
    ui->setupUi(this);

    // 1. åŸºæœ¬ UI é€£ç·š
    connect(ui->horizontalSlider, &QSlider::valueChanged, this, &DinnerSelection::snapSliderToStep);
    connect(ui->btnPlus, &QPushButton::clicked, this, &DinnerSelection::increasePrice);
    connect(ui->btnMinus, &QPushButton::clicked, this, &DinnerSelection::decreasePrice);

    // 2. è©•åˆ†è¤‡é¸æ¡†äº’æ–¥é‚è¼¯
    connect(ui->checkBox, &QCheckBox::clicked, [=] { ui->checkBox_2->setChecked(false); ui->checkBox_3->setChecked(false); });
    connect(ui->checkBox_2, &QCheckBox::clicked, [=] { ui->checkBox->setChecked(false); ui->checkBox_3->setChecked(false); });
    connect(ui->checkBox_3, &QCheckBox::clicked, [=] { ui->checkBox->setChecked(false); ui->checkBox_2->setChecked(false); });

    // 3. è·é›¢é€£ç·š
    connect(ui->sliderDistance, &QSlider::valueChanged, this, &DinnerSelection::onDistanceChanged);
    connect(ui->pushButton_4, &QPushButton::clicked, this, [=] { ui->sliderDistance->setValue(ui->sliderDistance->value() + 1); });
    connect(ui->pushButton_5, &QPushButton::clicked, this, [=] { ui->sliderDistance->setValue(ui->sliderDistance->value() - 1); });

    // 4. æŒ‰éˆ•åŠŸèƒ½é€£ç·š
    connect(ui->pushButton, &QPushButton::clicked, this, &DinnerSelection::applyFiltersAndShow);

    if (ui->btngood) {
        connect(ui->btngood, &QPushButton::clicked, this, [=](){
            int currentRow = ui->listRestaurant->currentRow();
            if (currentRow < 0 || currentRow >= currentFilteredRestaurants.size()) {
                QMessageBox::warning(this, "æç¤º", "è«‹å…ˆé¸æ“‡é¤å»³");
                return;
            }
            QJsonObject picked = currentFilteredRestaurants[currentRow];
            QMessageBox::information(this, "æˆåŠŸ", picked["name"].toString() + " å·²åŠ å…¥å–œå¥½ï¼");
        });
    }

    // éš¨æ©Ÿé¸å–æŒ‰éˆ•
    connect(ui->btnPick, &QPushButton::clicked, this, [=]() {
        if (currentFilteredRestaurants.isEmpty()) {
            ui->labelRandomResult->setText("ğŸ² éš¨æ©Ÿé¸å–\nè«‹å…ˆé€²è¡Œç¯©é¸");
            return;
        }

        int randomIndex = QRandomGenerator::global()->bounded(currentFilteredRestaurants.size());
        QJsonObject picked = currentFilteredRestaurants[randomIndex];

        QString name = picked["name"].toString();
        QJsonObject locObj = picked["geometry"].toObject()["location"].toObject();
        double lat = locObj["lat"].toDouble();
        double lon = locObj["lng"].toDouble();

        QObject *rootObj = mapWidget->rootObject();
        if (rootObj) {
            QMetaObject::invokeMethod(rootObj, "updateMapMarker",
                                      Q_ARG(QVariant, lat), Q_ARG(QVariant, lon), Q_ARG(QVariant, name));
        }

        double rating = picked["rating"].toDouble(-1);
        int level = picked["price_level"].toInt(-1);

        // åƒ¹æ ¼é¡¯ç¤ºé‚è¼¯
        QString priceRange = picked.contains("custom_price_text") ? picked["custom_price_text"].toString() :
                                 (level == -1 ? "ä¸€èˆ¬åƒ¹ä½" : (level == 0 ? "100å…§" : (level == 1 ? "100~200" : (level == 2 ? "200~300" : (level == 3 ? "300~500" : "500ä»¥ä¸Š")))));

        double dLat = (lat - 23.7019) * 111.0;
        double dLon = (lon - 120.4307) * 111.0 * cos(23.7019 * M_PI / 180.0);
        double dist = sqrt(dLat * dLat + dLon * dLon);

        ui->labelRandomResult->setText(
            QString("ğŸ² éš¨æ©Ÿçµæœï¼š\nåº—åï¼š%1\nâ­ è©•åˆ†ï¼š%2\nğŸ’° åƒ¹ä½ï¼š%3\nğŸ“ è·é›¢ï¼š%4 km")
                .arg(name).arg(rating < 0 ? "ç„¡" : QString::number(rating))
                .arg(priceRange).arg(QString::number(dist, 'f', 2))
            );
    });

    // æ–°å¢åº—å®¶æŒ‰éˆ•
    connect(ui->btnAdd, &QPushButton::clicked, this, [=]() {
        QObject *rootObj = mapWidget->rootObject();
        if (!rootObj) return;

        QDialog dialog(this);
        dialog.setWindowTitle("æ–°å¢ä¸­å¿ƒé»åº—å®¶");
        QFormLayout form(&dialog);
        QLineEdit *nameEdit = new QLineEdit(&dialog);
        QLineEdit *priceEdit = new QLineEdit(&dialog);
        QDialogButtonBox bb(QDialogButtonBox::Ok | QDialogButtonBox::Cancel, &dialog);
        form.addRow("åº—å:", nameEdit); form.addRow("åƒ¹ä½:", priceEdit); form.addRow(&bb);
        connect(&bb, &QDialogButtonBox::accepted, &dialog, &QDialog::accept);
        connect(&bb, &QDialogButtonBox::rejected, &dialog, &QDialog::reject);

        if (dialog.exec() == QDialog::Accepted && !nameEdit->text().isEmpty()) {
            QJsonObject ns;
            ns["name"] = nameEdit->text();
            ns["custom_price_text"] = priceEdit->text();
            ns["rating"] = 5.0;
            QJsonObject loc; loc["lat"] = rootObj->property("centerLat").toDouble(); loc["lng"] = rootObj->property("centerLng").toDouble();
            QJsonObject geo; geo["location"] = loc; ns["geometry"] = geo;
            allRestaurants.append(ns);
            applyFiltersAndShow();
        }
    });

    // 5. åˆå§‹åŒ–åœ°åœ–çµ„ä»¶
    mapWidget = new QQuickWidget(this);
    mapWidget->setResizeMode(QQuickWidget::SizeRootObjectToView);
    mapWidget->setSource(QUrl("qrc:/map.qml"));
    ui->mapLayout->addWidget(mapWidget);
    ui->labelMap->hide();

    ui->listRestaurant->setStyleSheet(
        "QListWidget::item { border-bottom: 1px solid #C0C0C0; padding: 8px; }"
        "QListWidget::item:selected { background-color: #e0f0ff; color: black; }"
        );

    connect(ui->listRestaurant, &QListWidget::itemClicked, this, [=]() {
        int currentRow = ui->listRestaurant->currentRow();
        if (currentRow >= 0 && currentRow < currentFilteredRestaurants.size()) {
            QJsonObject picked = currentFilteredRestaurants[currentRow];
            QJsonObject loc = picked["geometry"].toObject()["location"].toObject();
            QObject *rootObj = mapWidget->rootObject();
            if (rootObj) {
                QMetaObject::invokeMethod(rootObj, "updateMapMarker",
                                          Q_ARG(QVariant, loc["lat"].toDouble()),
                                          Q_ARG(QVariant, loc["lng"].toDouble()),
                                          Q_ARG(QVariant, picked["name"].toString()));
            }
        }
    });

    network = new QNetworkAccessManager(this);
    connect(network, &QNetworkAccessManager::finished, this, &DinnerSelection::onPlacesReply);

    fetchPlaces(23.7019, 120.4307);
}

DinnerSelection::~DinnerSelection() { delete ui; }

void DinnerSelection::snapSliderToStep(int value) {
    int snapped = (value + 50) / 100 * 100;
    if (snapped != value) ui->horizontalSlider->setValue(snapped);
    ui->label_3->setText(QString("%1 å…ƒ").arg(snapped));
}

void DinnerSelection::increasePrice() { ui->horizontalSlider->setValue(qMin(ui->horizontalSlider->value() + 100, ui->horizontalSlider->maximum())); }
void DinnerSelection::decreasePrice() { ui->horizontalSlider->setValue(qMax(ui->horizontalSlider->value() - 100, ui->horizontalSlider->minimum())); }
void DinnerSelection::onDistanceChanged(int value) { maxDistanceKm = value; ui->labelDistanceValue->setText(QString("%1 å…¬é‡Œå…§").arg(value)); }

void DinnerSelection::fetchPlaces(double lat, double lon, QString pageToken)
{
    QUrl url("https://maps.googleapis.com/maps/api/place/nearbysearch/json");
    QUrlQuery query;
    query.addQueryItem("key", "AIzaSyBs73o60jvr_scDSieQsGLJCIUhKmoBoOw");

    // åªæœ‰åœ¨ç¬¬ä¸€æ¬¡è«‹æ±‚æ™‚æ‰åˆå§‹åŒ–åˆ—è¡¨
    if (pageToken.isEmpty()) {
        allRestaurants.clear();
    }

    if (!pageToken.isEmpty()) {
        query.addQueryItem("pagetoken", pageToken);
    } else {
        query.addQueryItem("location", QString("%1,%2").arg(lat).arg(lon));
        query.addQueryItem("radius", "5000"); // æ“´å¤§æœå°‹åˆ° 5 å…¬é‡Œï¼Œç²å–æ›´å¤šæ¨£æœ¬
        query.addQueryItem("type", "restaurant");
        query.addQueryItem("language", "zh-TW");
    }

    url.setQuery(query);
    network->get(QNetworkRequest(url));
}


void DinnerSelection::onPlacesReply(QNetworkReply *reply)
{
    if (reply->error() != QNetworkReply::NoError) {
        reply->deleteLater();
        return;
    }

    QJsonObject root = QJsonDocument::fromJson(reply->readAll()).object();
    QJsonArray results = root["results"].toArray();

    // 1. ç›´æ¥é¡¯ç¤ºæŠ“åˆ°çš„æ‰€æœ‰åº—å®¶
    for (const QJsonValue &v : results) {
        QJsonObject obj = v.toObject();

        // æª¢æŸ¥é‡è¤‡
        bool exists = false;
        for (const QJsonValue &existingVal : allRestaurants) {
            if (existingVal.toObject()["place_id"].toString() == obj["place_id"].toString()) {
                exists = true;
                break;
            }
        }

        if (!exists) {
            allRestaurants.append(obj);
            currentFilteredRestaurants.append(obj); // åˆå§‹æ¸…å–®åŒ…å«æ‰€æœ‰æŠ“åˆ°çš„åº—
            addRestaurantToUI(obj);
        }
    }

    // 2. è™•ç†ä¸‹ä¸€é  Token
    m_nextPageToken = root["next_page_token"].toString();

    if (!m_nextPageToken.isEmpty()) {
        // å»¶é² 2 ç§’å†æŠ“ä¸‹ä¸€é ï¼Œé¿å… Google æ‹’çµ•è«‹æ±‚
        QTimer::singleShot(2000, this, [=]() {
            fetchPlaces(23.7019, 120.4307, m_nextPageToken);
        });
    }

    reply->deleteLater();
}

void DinnerSelection::addRestaurantToUI(const QJsonObject &obj)
{
    QString name = obj["name"].toString();
    double rating = obj["rating"].toDouble(-1);
    int priceLevel = obj["price_level"].toInt(-1);

    // è¨ˆç®—è·é›¢
    QJsonObject loc = obj["geometry"].toObject()["location"].toObject();
    double lat = loc["lat"].toDouble();
    double lng = loc["lng"].toDouble();

    double dLat = (lat - 23.7019) * 111.0;
    double dLon = (lng - 120.4307) * 111.0 * cos(23.7019 * M_PI / 180.0);
    double dist = sqrt(dLat * dLat + dLon * dLon);

    // åƒ¹ä½æ–‡å­—è½‰æ›
    QString priceRange = obj.contains("custom_price_text") ? obj["custom_price_text"].toString() :
                             (priceLevel == -1 ? "ä¸€èˆ¬åƒ¹ä½" :
                                  (priceLevel == 0 ? "100å…§" :
                                       (priceLevel == 1 ? "100~200" :
                                            (priceLevel == 2 ? "200~300" :
                                                 (priceLevel == 3 ? "300~500" : "500ä»¥ä¸Š")))));

    ui->listRestaurant->addItem(
        QString("ğŸ½ %1\n ğŸ’° %2\nâ­ %3\nğŸ“ %4 km")
            .arg(name).arg(priceRange)
            .arg(rating < 0 ? "ç„¡" : QString::number(rating))
            .arg(QString::number(dist, 'f', 2))
        );
}

void DinnerSelection::applyFiltersAndShow() {
    ui->listRestaurant->clear();
    currentFilteredRestaurants.clear();

    // å–å¾—ç¯©é¸é–€æª»
    double minRatingThreshold = 0.0;
    if (ui->checkBox_3->isChecked()) minRatingThreshold = 4.5;
    else if (ui->checkBox_2->isChecked()) minRatingThreshold = 4.0;
    else if (ui->checkBox->isChecked()) minRatingThreshold = 3.5;

    int sliderValue = ui->horizontalSlider->value();
    int maxPriceLevel = (sliderValue / 100) - 1;
    if (sliderValue >= 500) maxPriceLevel = 4;

    for (const QJsonValue &value : allRestaurants) {
        QJsonObject obj = value.toObject(); // ä¿®æ­£ï¼šå…ˆè½‰ç‚º Object å†é€²è¡Œæ“ä½œ
        double rating = obj["rating"].toDouble(-1);
        int priceLevel = obj["price_level"].toInt(-1);

        // è¨ˆç®—è·é›¢
        QJsonObject loc = obj["geometry"].toObject()["location"].toObject();
        double dLat = (loc["lat"].toDouble() - 23.7019) * 111.0;
        double dLon = (loc["lng"].toDouble() - 120.4307) * 111.0 * cos(23.7019 * M_PI / 180.0);
        double dist = sqrt(dLat * dLat + dLon * dLon);

        // åŸ·è¡Œç¯©é¸åˆ¤æ–·
        if (minRatingThreshold > 0 && rating < minRatingThreshold) continue;
        if (sliderValue > 0 && priceLevel != -1 && priceLevel > maxPriceLevel) continue;
        if (dist > maxDistanceKm) continue;

        // ç¬¦åˆæ¢ä»¶è€…
        currentFilteredRestaurants.append(obj);
        addRestaurantToUI(obj);
    }

    if (currentFilteredRestaurants.isEmpty()) {
        ui->label->setText("âœ¨ æ¯æ—¥æ¨è–¦ï¼š\nç›®å‰ç„¡ç¬¦åˆæ¢ä»¶çš„åº—å®¶");
    } else {
        int idx = QRandomGenerator::global()->bounded(currentFilteredRestaurants.size());
        QJsonObject dp = currentFilteredRestaurants[idx];
        QString dpPrice = dp.contains("custom_price_text") ? dp["custom_price_text"].toString() : "ä¸€èˆ¬åƒ¹ä½";
        QJsonObject loc = dp["geometry"].toObject()["location"].toObject();
        double dist = sqrt(pow((loc["lat"].toDouble()-23.7019)*111.0,2)+pow((loc["lng"].toDouble()-120.4307)*111.0*cos(23.7019*M_PI/180.0),2));

        ui->label->setText(
            QString("âœ¨ æ¯æ—¥æ¨è–¦ï¼š\nåº—åï¼š%1\nè©•åˆ†ï¼šâ­ %2\nåƒ¹ä½ï¼šğŸ’° %3\nè·é›¢ï¼šğŸ“ %4 km")
                .arg(dp["name"].toString())
                .arg(dp["rating"].toDouble(-1) < 0 ? "ç„¡" : QString::number(dp["rating"].toDouble()))
                .arg(dpPrice).arg(QString::number(dist, 'f', 2))
            );
    }
}
